sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -216,14 +216,36 @@ class Analyzer(
         GroupingSets(bitmasks(c), groupByExprs, child, aggregateExpressions)
       case Aggregate(Seq(r @ Rollup(groupByExprs)), aggregateExpressions, child) =>
         GroupingSets(bitmasks(r), groupByExprs, child, aggregateExpressions)
-      case x: GroupingSets =>
+      // Ensure all the expressions have been resolved.
+      case g: GroupingSets if g.expressions.exists(!_.resolved) =>
         val gid = AttributeReference(VirtualColumn.groupingIdName, IntegerType, false)()
-
+        // If users manually specify grouping__id in the aggregation expression, resolve it.
+        val aggExprs = g.aggregations.map(_.transform {
+          case u: UnresolvedAttribute if resolver(u.name, VirtualColumn.groupingIdName) => gid
+        }.asInstanceOf[NamedExpression])
+        if (aggExprs != g.aggregations) {
+          g.copy(aggregations = aggExprs, groupByExprs = g.groupByExprs :+ gid)
+        }
+        else {
+          g
+        }
+      case x: GroupingSets =>
```

Comment:
Add `if g.expressions.forall(_.resolved)` to make sure that all the expression are all resolved.